src/app/static/src/app/apiService.ts

```diff
@@ -0,0 +1,165 @@
+import axios, {AxiosError, AxiosResponse} from "axios";
+import {ActionContext, Commit} from "vuex";
+import {freezer, isMINTResponseFailure} from "./utils";
+import {RootState} from "./store";
+import {ResponseSuccess, ResponseFailure} from "./generated";
+import {RootMutation} from "./mutations";
+
+declare var appUrl: string;
+
+export interface ResponseWithType<T> extends Response {
+    data: T
+}
+
+export interface APIError {
+    error: string;
+    detail: string | null;
+}
+
+export interface API<S, E> {
+
+    withError: (type: E) => API<S, E>
+    withSuccess: (type: S) => API<S, E>
+    ignoreErrors: () => API<S, E>
+
+    postAndReturn<T>(url: string, data: any): Promise<void | ResponseWithType<T>>
+
+    get<T>(url: string): Promise<void | ResponseWithType<T>>
+
+    delete(url: string): Promise<void | true>
+}
+
+export class APIService<S extends string, E extends string> implements API<S, E> {
+
+    private readonly _commit: Commit;
+    private readonly _headers: any;
+
+    constructor(context: ActionContext<RootState, RootState>) {
+        this._commit = context.commit;
+    }
+
+    // appUrl will be set as a jest global during testing
+    private readonly _baseUrl = typeof appUrl !== "undefined" ? appUrl : "";
+
+    private _buildFullUrl = (url: string) => {
+        return this._baseUrl + url
+    };
+
+    private _ignoreErrors: Boolean = false;
+    private _freezeResponse: Boolean = false;
+
+    static getFirstErrorFromFailure = (failure: ResponseFailure) => {
+        if (failure.errors.length == 0) {
+            return APIService
+                .createError("API response failed but did not contain any error information. Please contact support.");
+        }
+        return failure.errors[0];
+    };
+
+    static createError(detail: string) {
+        return {
+            error: "MALFORMED_RESPONSE",
+            detail: detail
+        }
+    }
+
+    private _onError: ((failure: ResponseFailure) => void) | null = null;
+
+    private _onSuccess: ((success: ResponseSuccess) => void) | null = null;
+
+    freezeResponse = () => {
+        this._freezeResponse = true;
+        return this;
+    };
+
+    withError = (type: E) => {
+        this._onError = (failure: ResponseFailure) => {
+            this._commit(type, APIService.getFirstErrorFromFailure(failure));
```

Comment:
I've also switched the commit syntax to this https://vuex.vuejs.org/guide/mutations.html#commit-with-payload
away from https://vuex.vuejs.org/guide/mutations.html#object-style-commit because the latter is one of the things that felt like it added extra boilerplate in hint.